### ISSUE TITLE
Using external JSON de/serialization library

### ISSUE THREAD

**Issue by BottledByte** [tuomount/Open-Realms-of-Stars#665]
The game's code contains custom JSON parser. While it gets job done, it's functionality is very basic. It cannot serialize objects, it cannot deserialize into objects, it lacks many quality-of-life features and it is additional code that has to be maintained.

Adoption external library for working with JSON format would fixed all of the above mentioned issues, with addition that programmers might know the library from other projects. So it could be more comfortable and easier using it instead of the custom solution. And it will provide new options for working with game data.

As the game will unlikely ever deal with JSON and/or other formats excessively, using smaller de/serialization library, like `jackson-jr`, should be IMHO sufficient.

**Comment by tuomount** [comment]
Reason why it uses my own JSON library is that Jackson is with Apache license which is not compatible with GPL2. I don't know if there is GPL2 JSON parser. If that's the case that library could be separated as own project.

**Comment by tuomount** [comment]
Currently JSONs are only used with HUE lamps. I don't see big benefit changing current JSON parser to new library. Serialization is easy for creating objects, I understand that, but almost every big Java vulnerabilities have been in serialization since it  may allow executing code. For example Log4J was serialization issue, Jackson has had couple of serialization vulnerabilities. I am sure custom parser does not have those since, it does not support serialization.

Also I have fuzz tested couple years ago and fixed those found issues. So I my guess is it is pretty robust for what it is currently doing.

As I see Jackson is the JSON parser Java projects commonly use, I am not sure how common JSON-java is, so it might not it is not that common to other programmers. What game data where you thinking to get into JSON format? Also for end users changing the library does not bring any benefits, so I am not that keen to change it unless there is like advances, like storing some game data in JSON format. 

For example it could be used for save files, it probably would increase the save file size, and might even slowdown the loading of save files. Also that library seems to support serialization of Java objects to JSON but not back from JSON to java objects.

**Comment by tuomount** [comment]
Those gigantic factory classes in some other format would make sense, but is JSON good for it? JSON does not support comments, where as XML would. But neither of those aren't that fun to write either, but it would making modification maybe easier. At least I would prefer writing XML instead of JSON by hand, XML is a bit more readable, but that is just my view. I cannot remember out of my head what contained each factory, would some simple one line for each component/item etc work. If each line would contain all the parameters. 


Do you know where is the source code for that JSON-in-Java library? I checked the maven repository and it looks like it has two high vulnerabilities each year: https://mvnrepository.com/artifact/org.json/json

Vulnerabilities are not bad, since most user probably do use it without the HUE lamps and using those also requires manual configuration has some kind of certificate pinning, which would require to know at least which kind of ID that used HUE bridge has. But as my day job I work with vulnerabilities and I would like to do something else on my free time :smile:

**Comment by tuomount** [comment]
> CSV cannot represent hierarchical data easily, which makes it impractical.

This is absolutely true. So CSV is not good choice for this.
So Since discussion is moved to #666 I think this can be closed then?

**Comment by tuomount** [comment]
> If I understood final conclusion made in https://github.com/tuomount/Open-Realms-of-Stars/issues/666 correctly, JSON-java aka org.json aka JSON-in-Java library will be probably used instead of current in-house parser.
Yes

Seems it fails to parse empty array. It always expect that array has something inside. Obvious bug.

**Comment by tuomount** [comment]
Are you sure about that `javax.security.cert.X509Certificate` class. BlindTrustManager class uses `java.security.cert.X509Certificate` which suppose to be still valid in Java21.

joinGroup was easy fix, when I found how to make it with old usage without need to choose interface.

**Comment by tuomount** [comment]
This parser is somewhat better(less boilerplate, after using utility) than my JSON parser but also annoying. Methods do not generally return null, instead they throw exceptions which kind of annoying, but as a work around I made utility class that catches exception and returns null.
But then there is issue with HUE registration. HUE Bridge returns JSON array if there is an error and JSONObject if something is successful. There is no method to parse what ever JSON available. In my parser there was this JsonRoot which could handle what ever JSON type at beginning. 

But if there are no errors from bridge lights are working. It is just the registeration part that cause now the issue. I'll continue with this tomorrow.

**Comment by tuomount** [comment]
Ah, optXXXX(key) works as my utility. I looked example how to use this API and it used getters. Currently there is still problem that optIntegerObject works quite interesting. It returns value 0 if value is not present in JSON. This does not work, since values can be between 0-255 or 0-65535 in hue lamps. If lamp is not color lamp it is missing the hue value totally, but this optIntegerObject returns zero. I need to read more about javadocs, I hope I don't need to use tokener.

Also what I noticed that there is no method that would go through full json to search certain key and pick first one it encountered. Instead I needed to find each level and then look key from there and then next and so on. This is because each level is hashmap and opt and get only fetches keys from that level. That also caused some pondering why error handling was not working.
